### PR TITLE
MAYA-122104 ProxyShapeSceneSegmentHandler and UsdCameraHandler::findCameras.

### PR DESCRIPTION
**Pull request by williamkrick**

### CODE REVIEW

File: lib/mayaUsd/ufe/ProxyShapeSceneSegmentHandler.cpp
Comment (williamkrick) [inline lib/mayaUsd/ufe/ProxyShapeSceneSegmentHandler.cpp:56]: This extra `proxyShapePath == path` block is what is necessary to handle the special case of going into the child scene segments when passed a gateway item.

File: lib/mayaUsd/ufe/UsdCameraHandler.cpp
Comment (williamkrick) [inline lib/mayaUsd/ufe/UsdCameraHandler.cpp:60]: Native USD search for cameras.

**Review by vlasovi — DISMISSED** [review]
Looks good. I'm only wondering how much faster is this approach compared to just traversing Ufe::Hierarchy. In the end we still have to traverse USD hierarchy in search of cameras.

**Comment by williamkrick** [comment]
> Looks good. I'm only wondering how much faster is this approach compared to just traversing Ufe::Hierarchy. In the end we still have to traverse USD hierarchy in search of cameras.

You are right in that the complexity of the algorithm is the same. The speed difference comes from eliminating the overhead of calls to Ufe during the traversal. Using Ufe::Hierarchy to traverse means creating many additional Ufe::SceneItems, Ufe::Paths and Ufe::Hierarchy objects, which means allocating a bunch of std::vector<>, std::shared_ptr<> etc. Doing the iteration with USD means we don't do any allocation when the object is not a gateway item or a camera (which is most objects).

Avoiding the overhead of Ufe when doing the hierarchy traversal ends up saving a significant amount of time. This version is something like 10x faster then the original code.

File: lib/mayaUsd/ufe/Global.cpp
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/Global.cpp:276]: Shouldn't we reset the g_MayaCameraHandler as well?

File: lib/mayaUsd/ufe/ProxyShapeCameraHandler.cpp
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/ProxyShapeCameraHandler.cpp:55]: Do we really need to do this?  The ProxyShape is a shape node --- shape nodes only have children in limited circumstances, and can they really be cameras?

File: lib/mayaUsd/ufe/ProxyShapeCameraHandler.cpp
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/ProxyShapeCameraHandler.cpp:56]: Shouldn't this be the else clause of the isAGatewayType()?  E.g. as per
https://github.com/Autodesk/maya-usd/blob/3bd67035ccee8936b68a2623e9defe0551204c0a/lib/mayaUsd/ufe/ProxyShapeHierarchyHandler.cpp#L50

File: lib/mayaUsd/ufe/ProxyShapeCameraHandler.cpp
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/ProxyShapeCameraHandler.cpp:65]: If my hunch above is right, you can just return UsdCameraHandler::find(...);

File: lib/mayaUsd/ufe/ProxyShapeSceneSegmentHandler.cpp
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/ProxyShapeSceneSegmentHandler.cpp:84]: Couldn't this just be
return stage ? true : fMayaSceneSegmentHandler ? fMayaSceneSegmentHandler->isGateway_(path) : false;

File: lib/mayaUsd/ufe/ProxyShapeSceneSegmentHandler.h
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/ProxyShapeSceneSegmentHandler.h:19]: Do we need this include?

File: lib/mayaUsd/ufe/UsdCameraHandler.cpp
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/UsdCameraHandler.cpp:77]: Nice!

**Review by ppt-adsk — CHANGES_REQUESTED** [review]
Needs a test.

**Comment by williamkrick** [comment]
These changes are already merged on top of @JGamache-autodesk work for https://github.com/Autodesk/maya-usd/pull/2315, so merge that one first.

The only thing I'm missing is a test for SceneSegmentHandler. I forgot the Ufe::RunTimeMgr Python API that I need to write the test, so it is going to come in another PR.

**Comment by seando-adsk** [comment]
I added the label "do not merge yet" as a reminder to myself to merge #2315 first.

File: test/lib/ufe/testSceneSegment.py
Comment (ppt-adsk) [inline test/lib/ufe/testSceneSegment.py:57]: Why is this different from setUp()?

**Review by ppt-adsk — CHANGES_REQUESTED** [review]
Thanks for the tests.  Would be great if you could adjust the test conditionals to follow the pattern we've done in the past, makes it easier to remove these obsolete conditionals at the next Maya preview release.

Was quite confusing to review your changes and Jerry's changes in the same pull request.  Can't we just merge in Jerry's changes first, then you can update from dev and rebase your commits on top?

File: lib/mayaUsd/ufe/CMakeLists.txt
Comment (williamkrick) [reply lib/mayaUsd/ufe/CMakeLists.txt:114]: We're in kind of a strange state because we've had a bunch of UFE preview versions go by without getting them into Maya. Everyone is building locally slightly differently. But no QA-testable build with Ufe 0.4.9 is ever going to exist, which means these could all be testing for 0.4.10 (or 0.4.11 if my extra python API changes are merged). That simplifies the cmake. But we don't know which version to check for until something gets merged into Maya, and we can't merge into Maya until this gets merged...

Comment (ppt-adsk) [reply lib/mayaUsd/ufe/CMakeLists.txt:114]: It seems to me we can merge this with GREATER_EQUAL 4010 (or 4011) at any time.  Local developer builds will have 0.4.10 (or 0.4.11, as the case may be), so maya-usd will work with the newer version, as well as with the older UFE version.  Am I missing something?

File: test/lib/ufe/testCamera.py
Comment (williamkrick) [reply test/lib/ufe/testCamera.py:284]: The Maya camera handler is not implemented, and there are no cameras that are children of proxyShapeParentItem. If we had the Maya camera handler and we had searched the Maya root path then we'd find front, side, top and persp.

Comment (ppt-adsk) [reply test/lib/ufe/testCamera.py:284]: Right, so do we plan on leaving the Maya camera handler unimplemented, because there is no demonstrated need for it?  If so, would be nice to add a comment to that effect in the test.

File: test/lib/ufe/testSceneSegment.py
Comment (williamkrick) [reply test/lib/ufe/testSceneSegment.py:57]: I'm not really sure. I copied most of the test from testCamera.py and it has this pattern too. Probably I copied it for testCamera from somewhere else.

I guess if all the tests are running with the same USD file (which for testCamera and testSceneSegment they are running with the same USD file, then the file could get loaded in setUp and each test could re-use it? That would make the test run faster. That should work relatively easily here because I'm not actually modifying the file at all.

I'll try it and see how it goes.

File: test/lib/ufe/testCamera.py
Comment (williamkrick) [reply test/lib/ufe/testCamera.py:284]: Sure I'll add a comment.

**Comment by williamkrick** [comment]
The new changes were around the feedback provided by @ppt-adsk on the tests for CameraHandler and SceneSegementHandler. I also re-based on top of dev to eliminate merge conflicts.

File: test/lib/ufe/testCamera.py
Comment (ppt-adsk) [inline test/lib/ufe/testCamera.py:250]: I wonder about these single-use UFE path segments --- can't we just use ufe.PathString.path() to create them out of concatenated strings?  E.g.
mayaPathSegmentStr = '|stage|stageShape'
camerasParentPath = ufe.PathString.path(mayaPathSegmentStr+'/cameras/cam2')
I get that you're re-using the Maya path segment, and in real code there may be a performance benefit to doing this, but in a test it seems so verbose.

File: test/lib/ufe/testCamera.py
Comment (ppt-adsk) [inline test/lib/ufe/testCamera.py:263]: Have a look at ufeUtils.createItem().  It will create a ufe.SceneItem directly out of a path string, I think this is what you want, and will remove a lot of boiler-plate code from your test.

**Review by ppt-adsk — COMMENTED** [review]
Not a big deal, but it feels like the tests could have some fat trimmed off, sorry for noticing this late.  Otherwise looks great!

**Comment by williamkrick** [comment]
Updated the tests to create the required paths in a more concise way. Updated the Ufe preview version checks to the latest.

**Review by ppt-adsk — APPROVED** [review]

**Pull request merged**
